### PR TITLE
Update OpenSSL to version 3.0.8 in the base system

### PR DESCRIPTION
**Pull request by khorben**
This branch is the closest I am currently to a functional update to OpenSSL 3.0.8 in FreeBSD's base system. It was started from the existing `vendor/openssl-3.0` branch.
The thorough review expected should include:
* The configuration applied (default from the corresponding `security/openssl30` port with the legacy provider enabled)
* Security patches up to date (the port fixes CVEs, I have not included them yet)
* Other patches in the corresponding port may be relevant too (I have not included them)
* API changes reflected accurately in `secure/lib/libcrypto/Version.map` and `secure/lib/libssl/Version.map`
* Whether the choice of `SHLIB_MAJOR` is good (currently `30` since `3` is already obsolete)
* Functional tests on Tier-1 architectures (amd64, aarch64...)
* Relevance of the `ossl-modules` providers
* Patches to software using OpenSSL are correct (especially libradius)
* The included update of ldns is correct (from the `vendor/ldns-1.8.3` tag by YT)
* This does not touch the OpenSSL files that were moved into `sys/crypto/openssl` to avoid any trouble with the kernel, and re-imports them into `crypto/openssl` instead; security fixes beware of both locations

In most software users of OpenSSL, a compatibility compilation flag was used in order to expose and use the former OpenSSL API.

Sponsored by: FreeBSD Foundation

### CODE REVIEW

**Comment by ngie-eign** [comment]
@khorben : you're awesome! Would you be opposed to me taking bits of your work, cherry-picking/squashing them, then pushing bits and pieces into base?

**Comment by ngie-eign** [comment]
My brain cells have been under stimulated lately. I'd really like to do some straightforward FreeBSD development.

**Comment by ngie-eign** [comment]
Hmm.. the build failed with amd64. Let's see if I can track down why and propose a change in the PR.

**Comment by khorben** [comment]
> @khorben : you're awesome! Would you be opposed to me taking bits of your work, cherry-picking/squashing them, then pushing bits and pieces into base?

This is great to hear and absolutely, feel free to sort it out and clear the way :)
I tried to keep the commit history relatively clean, I just couldn't put all of the relevant changes directly in the merge commit.

**Comment by khorben** [comment]
> Hmm.. the build failed with amd64. Let's see if I can track down why and propose a change in the PR.

Classic, I forgot to `git add` a file; I have updated the branch.

**Comment by emaste** [comment]
@ngie-eign I think all of the `CFLAGS+=-DOPENSSL_API_COMPAT=0x10100000L` additions should be good to go now.

I used "specify OpenSSL 1.1 API" instead of "fix building with OpenSSL 3.0" in the commit message subjects, and these commits should have a `Sponsored by: The FreeBSD Foundation`.

**Comment by emaste** [comment]
Hmm, dma fails with `OPENSSL_API_COMPAT` set, with:

```
/usr/home/emaste/src/freebsd-git/main/contrib/dma/crypto.c:118:2: error: call to undeclared function 'SSL_library_init'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
        SSL_library_init();
        ^
/usr/home/emaste/src/freebsd-git/main/contrib/dma/crypto.c:119:2: error: call to undeclared function 'SSL_load_error_strings'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
        SSL_load_error_strings();
        ^
/usr/home/emaste/src/freebsd-git/main/contrib/dma/crypto.c:119:2: note: did you mean 'ERR_lib_error_string'?
/usr/include/openssl/err.h:239:13: note: 'ERR_lib_error_string' declared here
const char *ERR_lib_error_string(unsigned long e);
            ^
2 errors generated.
```

Indeed,
> The SSL_library_init() and OpenSSL_add_ssl_algorithms() functions were deprecated in OpenSSL 1.1.0 by OPENSSL_init_ssl().

so IMO we should update dma and add the `OPENSSL_API_COMPAT` #define first

File: lib/libfido2/Makefile [hunk 1]
```diff
@@ -68,6 +68,7 @@ CFLAGS+= -DTLS=__thread
 CFLAGS+= -D_FIDO_MAJOR=1
 CFLAGS+= -D_FIDO_MINOR=10
 CFLAGS+= -D_FIDO_PATCH=0
+CFLAGS+= -DOPENSSL_API_COMPAT=0x10100000L
```
[the remark this answers is not in the document]
Comment (ngie-eign) [reply lib/libfido2/Makefile]: Does updating libfido2 fix this issue?

**Comment by ngie-eign** [comment]
> Hmm, dma fails with `OPENSSL_API_COMPAT` set, with:
> 
> ```
> /usr/home/emaste/src/freebsd-git/main/contrib/dma/crypto.c:118:2: error: call to undeclared function 'SSL_library_init'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
>         SSL_library_init();
>         ^
> /usr/home/emaste/src/freebsd-git/main/contrib/dma/crypto.c:119:2: error: call to undeclared function 'SSL_load_error_strings'; ISO C99 and later do not support implicit function declarations [-Werror,-Wimplicit-function-declaration]
>         SSL_load_error_strings();
>         ^
> /usr/home/emaste/src/freebsd-git/main/contrib/dma/crypto.c:119:2: note: did you mean 'ERR_lib_error_string'?
> /usr/include/openssl/err.h:239:13: note: 'ERR_lib_error_string' declared here
> const char *ERR_lib_error_string(unsigned long e);
>             ^
> 2 errors generated.
> ```
> 
> Indeed,
> 
> > The SSL_library_init() and OpenSSL_add_ssl_algorithms() functions were deprecated in OpenSSL 1.1.0 by OPENSSL_init_ssl().
> 
> so IMO we should update dma and add the `OPENSSL_API_COMPAT` #define first

Approved internal change for merge with FreeBSD.
I provided a complete patch upstream to https://github.com/corecode/dma/pull/126 to leverage the OpenSSL 3 APIs if/when possible.

File: lib/libfido2/Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (emaste) [reply lib/libfido2/Makefile]: libfido2's upstream build does:
```
        if(CRYPTO_VERSION VERSION_GREATER_EQUAL 3.0)
                add_definitions(-DOPENSSL_API_COMPAT=0x10100000L)
        endif()
```
with a few special case `#if OPENSSL_VERSION_NUMBER >= 0x30000000` here and there, but I imagine a future update will address this.

**Review by ngie-eign — COMMENTED** [review]
Submitted upstream as https://github.com/libarchive/libarchive/pull/1869 .

File: lib/libarchive/Makefile.inc [hunk 2]
```diff
@@ -10,7 +10,7 @@ CFLAGS+=	-DPLATFORM_CONFIG_H=\"${.CURDIR}/config_freebsd.h\"
 
 .if ${MK_OPENSSL} != "no"
 CFLAGS+=	-DWITH_OPENSSL
-LIBADD+=	crypto
+LIBADD+=	ssl crypto
```
[the remark this answers is not in the document]
Comment (ngie-eign) [reply lib/libarchive/Makefile.inc]: Following up on my comment... libssl should not be required here. None of the libssl-related APIs are used in libarchive.
My guess is that this change was made based on similar code in the upstream pkg-config. I submitted https://github.com/libarchive/libarchive/pull/1870 to correct the overlinking.

**Review by ngie-eign — COMMENTED** [review]
Review: https://reviews.freebsd.org/D40082

**Review by ngie-eign — CHANGES_REQUESTED** [review]
This isn't right. libmd and libcrypto shouldn't be mixed.

**Comment by ngie-eign** [comment]
> This isn't right. libmd and libcrypto shouldn't be mixed.

To be clear (because my last statement wasn't potentially): the implementations *can* mix by design, but there's no reason why there should be 2 crypto implementations linked into a library.

This effectively overlinks libmd into libfetch when MK_OPENSSL != no.

**Review by ngie-eign — COMMENTED** [review]
The lib/ldns change was submitted as c6750ddec9177b67bb4883717933e87ac24a3a44.

**Comment by ngie-eign** [comment]
> > Review: https://reviews.freebsd.org/D40082
> 
> The lib/ldns change was submitted as c6750ddec9177b67bb4883717933e87ac24a3a44.

The above commit broke the build; I reverted it in 902dc544cc5859e24c727126e45416133c1c6d46.

This is something that will likely need to come in as part of the entire OpenSSL 3.0 deliverable, or adjusted to work with main, as shipped today.

**Comment by emaste** [comment]
> > > Review: https://reviews.freebsd.org/D40082
> > 
> > 
> > The lib/ldns change was submitted as [c6750dd](https://github.com/freebsd/freebsd-src/commit/c6750ddec9177b67bb4883717933e87ac24a3a44).
> 
> The above commit broke the build; I reverted it in [902dc54](https://github.com/freebsd/freebsd-src/commit/902dc544cc5859e24c727126e45416133c1c6d46).

This is perhaps the same sort of issue we encountered with dma, but a more involved example?

**Comment by khorben** [comment]
I have now marked this PR as "ready", with the following to keep in mind:
* The configuration applied is as per the `security/openssl30` port, with the legacy provider enabled
* This is strictly the 3.0.8 release, security patches issued since that release are ___not___ included; the CVE fixes found in the `security/openssl30` port are ___not___ included here (they probably deserve their own respective commits in order to keep track)
* Other patches in the corresponding port may be relevant too (I have not included them)
* I believe the API changes to be reflected accurately in `secure/lib/libcrypto/Version.map` and `secure/lib/libssl/Version.map`
* The choice of `SHLIB_MAJOR` is `3` in OpenSSL's header files (as per upstream) but `30` for the `.so` files (since `.3` has already been used by FreeBSD before in base)
* The following architectures are believed to build: amd64, aarch64, i386
* The `ossl-modules` providers have not been tested yet
* This does not touch the OpenSSL files that were moved in bc3d5698008e9b3b19495e853cbc2598979ccf8a into `sys/crypto/openssl` to avoid any trouble with the kernel, and re-imports them into `crypto/openssl` instead; security fixes beware of both locations
* The assembly files for amd64, i386, aarch64 were obtained by cross-compiling the original archive from source instead of with `Makefile.asm`

**Comment by emaste** [comment]
cc @bsdjhb

**Comment by khorben** [comment]
Unfortunately this pull-up request has to go back to draft after the update to 1.1.1u in `main`.

**Comment by khorben** [comment]
Closing this pull-up request now that the direct update to OpenSSL 3.0.9 is almost ready.

**Pull request closed without merging**